Data/ByteString/Short/Internal.hs

```diff
@@ -615,12 +615,29 @@ useAsCStringLen bs action =
 -- @since 0.11.3.0
 isValidUtf8 :: ShortByteString -> Bool
 isValidUtf8 sbs@(SBS ba#) = accursedUnutterablePerformIO $ do
-  i <- cIsValidUtf8 ba# (fromIntegral (length sbs))
+  -- Use a safe FFI call for large inputs to avoid GC synchronization pauses
+  -- in multithreaded contexts.
+  -- This specific limit was chosen based on results of a simple benchmark, see:
+  -- https://github.com/haskell/bytestring/issues/451#issuecomment-991879338
+  -- When changing this function, also consider changing the related function:
+  -- Data.ByteString.isValidUtf8
+  let n = length sbs
+  i <- if n < 1000000
+     then cIsValidUtf8 ba# (fromIntegral n)
+     else cIsValidUtf8Safe ba# (fromIntegral n)
```

Comment:
Also, is `ba#` guaranteed to be kept alive for the duration of this call? Or do we need some `touch#` or `with...` stuff here?

Reply:
I think you are right that some kind of touch is necessary here, although I don't know what would be the best way to do that.

Reply:
I think checking `isPinnedByteArray#` is sufficient here. We do not extract a pointer to `byteArrayContent#`, so it's GHC's responsibility not to garbage collect `ba#` before time.

Reply:
Ah, so the `touch#` only has to be used if you actually use a `Ptr`, not with plain byte arrays. @clyring do you agree that this doesn't need any further action?

Reply:
My quick testing does suggest that GHC currently does keep `ByteArray#` foreign function parameters alive for the duration of even a safe FFI call. But if this behavior is documented anywhere I wasn't able to find it. (I should raise a ticket on GHC's issue tracker about this.) I would personally prefer some `touch#` anyway to be safe since it should have no runtime cost and if it turns out to have been necessary in some situation the resulting bug will be near-impossible for a user to track down.

Reply:
I’m not sure what exactly you suggest to touch. byteArrayContent#?

Reply:
I had in mind the `ByteArray#` reference itself, `ba#`. Though I concede it's possible that this won't do what I expect.

But, now that I check, the quick test I hacked together also fails to detect the unsafety of `Addr#` passing without any form of `touch#` or `keepAlive#`. So my previous comment should not be seen as persuasive evidence.

Reply:
I do not follow. We have a thunk depending on `ba#` - how could it be garbage collected before the thunk is fully evaluated?

Reply:
Well, my concern is that there might not be a thunk depending on `ba#` at runtime, if `ba#` is not subsequently used in Haskell code. I don't know the exact inner workings of GHC's implementation of safe FFI calls at runtime. If it involves creating a pinned thunk that holds on to the arguments of the FFI call, then that would obviously keep alive the `ByteArray#` argument. But this isn't made clear by the documentation, and it's easy to imagine an implementation that doesn't provide such a guarantee.